src/ui/components/SecondaryToolbox/react-devtools/rdtProcessing.ts

```diff
@@ -0,0 +1,535 @@
+  nodeType: "root";
+  isStrictModeCompliant: boolean;
+  supportsProfiling: boolean;
+  supportsStrictMode: boolean;
+  hasOwnerMetadata: boolean;
+}
+
+interface TreeOperationAddNodeContents {
+  nodeType: "node";
+  parentId: number;
+  ownerId: number;
+  stringTableIndex: number;
+  keyStringTableIndex: number;
+}
+
+interface TreeOperationAddBase {
+  type: typeof TREE_OPERATION_ADD;
+  nodeId: number;
+  nodeType: number;
+}
+
+interface TreeOperationAddRoot extends TreeOperationAddBase {
+  name: "addRoot";
+  contents: TreeOperationAddRootContents;
+}
+
+interface TreeOperationAddNode extends TreeOperationAddBase {
+  name: "addNode";
+  contents: TreeOperationAddNodeContents;
+}
+
+interface TreeOperationRemove {
+  type: typeof TREE_OPERATION_REMOVE;
+  name: "remove";
+  nodeIds: number[];
+}
+
+interface TreeOperationRemoveRoot {
+  type: typeof TREE_OPERATION_REMOVE_ROOT;
+  name: "removeRoot";
+}
+
+interface TreeOperationSetSubtreeMode {
+  type: typeof TREE_OPERATION_SET_SUBTREE_MODE;
+  name: "setSubtreeMode";
+  rootId: number;
+  mode: number;
+}
+
+interface TreeOperationReorderChildren {
+  type: typeof TREE_OPERATION_REORDER_CHILDREN;
+  name: "reorderChildren";
+  nodeId: number;
+  children: number[];
+}
+
+interface TreeOperationUpdateTreeBaseDuration {
+  type: typeof TREE_OPERATION_UPDATE_TREE_BASE_DURATION;
+  name: "updateTreeBaseDuration";
+  id: number;
+  baseDuration: number;
+}
+
+interface TreeOperationUpdateErrorsOrWarnings {
+  type: typeof TREE_OPERATION_UPDATE_ERRORS_OR_WARNINGS;
+  name: "updateErrorsOrWarnings";
+  nodeId: number;
+  errors: number;
+  warnings: number;
+}
+
+export type ParsedReactDevtoolsTreeOperations =
+  | TreeOperationAddRoot
+  | TreeOperationAddNode
+  | TreeOperationRemove
+  | TreeOperationRemoveRoot
+  | TreeOperationSetSubtreeMode
+  | TreeOperationReorderChildren
+  | TreeOperationUpdateTreeBaseDuration
+  | TreeOperationUpdateErrorsOrWarnings;
+
+export function deconstructOperationsArray(originalOperations: number[]) {
+  const rendererId = originalOperations[0];
+  const rootId = originalOperations[1];
+
+  let i = 2;
+
+  // Reassemble the string table.
+  const stringTable: string[] = [
+    null as any, // ID = 0 corresponds to the null string.
+  ];
+
+  const stringTableSize = originalOperations[i++];
+  const stringTableEnd = i + stringTableSize;
+  while (i < stringTableEnd) {
+    const nextLength = originalOperations[i++];
+    const nextString = utfDecodeString(originalOperations.slice(i, i + nextLength));
+    stringTable.push(nextString);
+    i += nextLength;
+  }
+
+  const numericTreeOperations = originalOperations.slice(stringTableEnd);
+
+  const parsedTreeOperations = parseTreeOperations(numericTreeOperations);
+
+  return {
+    rendererId,
+    rootId,
+    stringTable,
+    treeOperations: parsedTreeOperations,
+  };
+}
+
+export function parseTreeOperations(treeOperations: number[]): ParsedReactDevtoolsTreeOperations[] {
```

Comment:
Functions like `parseTreeOperations ` and `reconstructOperationsArray` should have unit tests. Doesn't seem like it would require a lot of effort to make them, and a regression in functions like this could have a large negative impact.

I don't think it's realist to review operations array deconstruction+reconstruction code like this as part of a PR anyway. Mistakes would be way too subtle.

Reply:
Agreed, these are great candidates for unit tests.

Talked to Jason about this. We agreed that I'll merge this now, and write the tests first thing after I get back from PTO:

- https://linear.app/replay/issue/FE-1720/write-tests-for-rdt-routine-processing-logic